File: Simulated Annealing/anneal.py
import math 
import subprocess
import networkx as nx
import matplotlib.pyplot as plt
from random import *
from copy import deepcopy

class Annealer:
	
	def __init__(self, init_state, costfn, local_move, cool_func, disp=-1):
		self.costfn=costfn
		self.local_move=local_move
		self.cool_func=cool_func
		
		self.best_state=init_state
		self.best_cost=self.costfn(init_state)
		self.disp=disp
		

	def accept(self, delta, temp):
		if delta<0:
			return 1.0
		return math.exp(-delta/temp)
	
	def anneal(self, temp):
		
		current_state=self.best_state
		current_cost=self.best_cost

		iters=-1
		while(temp > 1):
			
			print "BEST: %f  T: %f   C: %f "%(self.best_cost, temp, current_cost)
			if self.disp>0 and iters%self.disp==0:
				display_G(current_state)
			iters+=1

			(state, delta)=self.local_move(current_state)
			
			if self.accept(delta, temp) > random():
				current_state=state
				current_cost+=delta
				
				if current_cost<self.best_cost:
					self.best_cost=current_cost
					self.best_state=state
						
			temp=self.cool_func(temp)


def display_G(state):
	plt.clf()
	nx.draw(nxG, nxPos, node_color=state)
	plt.pause(0.5)



#--------------Parsing is a bi*** apparently-------------#
N, E=map(int,raw_input().split())
G={}
for i in range(E):
	u, v=map(int, raw_input().split())
	if u in G:
		G[u].append(v)
	else:
		G[u]=[v]
		
	if v in G:
		G[v].append(u)
	else:
		G[v]=[u]
		

#--------------Make an NX graph for prettiness-----------#
nxG=nx.Graph()
for node in G:
	nxG.add_node(node)
	for nbh in G[node]:
		nxG.add_edge(node,nbh)
nxPos=nx.spring_layout(nxG)



#--------------YOU NEED TO EDIT THESE FUNCTIONS AND VALUES-------------#
colours=2
init_state=tuple([0]*N) #you can leave this, or make it a random state
init_temp=100


'''
Remember, ANY local move works. Even extremely trivial ones.
We're hoping you'll come up with a decently complex local move --
with a good balance of computational expensiveness and utility.
The more complex local moves will help the solution converge faster,
and that's what you should be aiming to do
'''
def local_move(state):
	#FILL IN YOUR CODE HERE- and return the new state and the delta value
        new_state=[0]*N
        new_state=list(new_state)
        graph=deepcopy(G)
        palette=[[i for i in range(1,colours+1)] for x in range(N)]
        i=0
        while (i<N):
                #new_state[i+1]=state[i]
                #new_state[i]=randint(1,colours)
                if len(palette[i])!=0:
                        new_state[i]= choice(palette[i])
                        for j in G[i]:
                                if new_state[i] in palette[j]:
                                        palette[j].remove(new_state[i])
                        i+=1
                else:
                        '''new_state[i-1]=0
                        for j in G[i-1]:
                                if new_state[i-1] not in palette[j]:
                                        palette[j].append(new_state[i-1])
                        i-=1'''
                        return state,0
        
        #new_state[0]=state[N-1]
        new_state=tuple(new_state)
        delta=cost_fn(new_state)-cost_fn(state)
        #print palette
	return new_state, delta

'''
Try out all the temperature functions you can think of- Slow, fast, sinusoidal.
No, not sinusoidal.
'''
def cool_func(temp):
	#FILL IN YOUR CODE HERE- and return the new temperature value
        new_temp=temp-.1
	return new_temp

'''
Find a good objective function to MINIMIZE.
Remember, your ideal, best, amazing solution needs to have the
lowest function value
'''
def cost_fn(state):
	#FILL IN YOUR CODE HERE- and return the value of the state
        cost=0
        for i in range(0,N):
             for j in G[i]:
                     if state[j]!=state[i]:
                             cost+=0
                     else:
                             cost+=1
	return cost + .1*len(set(state))

#---------------------------------------------------------------------#


'''
disp controls after how many iterations you want to see your graph
set it to a negative number if you don't want to see it at all
(you'll still see your final graph)
'''
gc=Annealer(init_state, cost_fn, local_move, cool_func, disp=-1)
gc.anneal(init_temp)
print gc.best_state
print gc.best_cost
print len(set(gc.best_state))
display_G(gc.best_state)
#Save a picture of your result, to show off to your friends
plt.savefig('result.png')
plt.pause(100)

File: Simulated Annealing/anneal.cpp
#include<bits/stdc++.h>
#include<chrono>
#include<random>
#include <stdlib.h>

using namespace std;
#define ull unsigned long long

unsigned seed = chrono::system_clock::now().time_since_epoch().count();
default_random_engine gen(seed);

unordered_map<int,vector<int>> g;
unsigned long long N,E,v1,v2,colours;
string gid;
vector<int> parse(string state){

	stringstream lineStream(state);
    vector<int> numbers;
	int num;
	while (lineStream >> num)
		numbers.push_back(num);
	return numbers;
}



void init(string &current,unsigned long long colours)
{
    uniform_int_distribution<int> distr(48,47+colours);
    for(unsigned i=0; i<current.length();i++)
        current[i]=distr(gen);
}



void display_G(string state, double timeout, int save){
	
	
	
	ostringstream sstream;
	for(int i=0; i<state.length(); i++){
		sstream<<(int(state.at(i))-'0')<<" ";
	}	
	
	ostringstream cmd;
	cmd<<"python show_graph.py "<<gid<<" \""<<sstream.str()<<"\" "<<timeout<<" "<<save;
	
	char command[100];
	sprintf (command, cmd.str().c_str());  
	system(command);
	
	/*uncomment the next line if you want to pause every visualization
	until you hit ENTER, to get a better look at the pretty graph*/
	//cin.get();
	
	

}
/*-----------------------------YOU NEED TO EDIT THESE FUNCTIONS----------------------------------*/

/*Find a good objective function to MINIMIZE.
Remember, your ideal, best, amazing solution needs to have the
lowest function value*/
unsigned long long cost_fn(string state)
{
	//FILL IN YOUR CODE HERE- and return the value of the state
	return cost;

}


/*Try out all the temperature functions you can think of- Slow, fast, sinusoidal.
No, not sinusoidal.*/
double schedule (double t)
{
	//FILL IN YOUR CODE HERE- and return the new temperature value
    return new_t;
}


/* Remember, ANY local move works. Even extremely trivial ones.
We're hoping you'll come up with a decently complex local move --
with a good balance of computational expensiveness and utility.
The more complex local moves will help the solution converge faster,
and that's what you should be aiming to do */
string move(string current)
{
	//FILL IN YOUR CODE HERE- and return the new state
    return next;
}

/*------------------------------------------------------------------------------------------*/


int main(int argc, char**argv)
{


	/* disp(the value 300 here) controls after how many iterations you want to see your graph
	set it to a negative number if you don't want to see it at all.
	(you'll still see your final graph)*/
	/*--------------YOU NEED TO EDIT THESE VALUES-------------*/
	int disp=300;
	double t= 0.0; //initial temperature
	/*-------------------------------------------------------*/
	
	
	ifstream fin;
    fin.open(argv[1]);
	
	int idx=string(argv[1]).find("gc_");
	gid=string(argv[1]).substr(idx+3);
	
    colours = atoll(argv[2]);
    fin>>N>>E;

    for(int i=0;i<E;i++)
    {
        fin>>v1>>v2;
        g[v1].push_back(v2);
		g[v2].push_back(v1);
    }
    
    string current(N,0);

    init(current,colours);
    string best_cost(current);
	
	int iters=-1;
    while(t>1)
    {   unsigned long long cst_fn = cost_fn(current);
        ull bst_cst = cost_fn(best_cost);
        t= schedule(t);
        string next = move(current);
		
        cout<<"BEST: "<<bst_cst<<"    T:"<<t<<"    C:"<<cst_fn<<endl;
		
		if(disp>0 && iters%disp==0)
			display_G(next, 0.5, 0);
		iters+=1;
		
		ull  nxt_cst= cost_fn(next);
        long long  del_e =nxt_cst-cst_fn;
        if(del_e <0)
        {
            current = next;
			
			if(nxt_cst<bst_cst){
				bst_cst = nxt_cst;
				best_cost=next;
			}
				
        }
        else
        {
            uniform_real_distribution<double> distr(0.0,1.0);
           long double val1 = distr(gen);
            long double val2 = (long double)1.0/exp(((long double)del_e)/(long double)t);

            if(val1<= val2)
                current=next;
        }

    }
	
	ostringstream sstream;
	for(int i=0; i<best_cost.length(); i++){
		sstream<<(int(best_cost.at(i))-'0')<<" ";
	}	
	
    cout<<current<<" "<<cost_fn(current)<<endl;
    cout<<"Best :" << sstream.str()<<" with cost:"<<cost_fn(best_cost)<<"\n\n\n\n";
	
	display_G(best_cost, 100, 1);
	
}
